cub/agent/agent_reduce.cuh

```diff
@@ -27,357 +27,418 @@
+                            (std::is_pointer<InputIteratorT>::value) &&
+                            Traits<InputT>::PRIMITIVE;
+
+  static constexpr CacheLoadModifier LOAD_MODIFIER =
+    AgentReducePolicy::LOAD_MODIFIER;
+
+  static constexpr BlockReduceAlgorithm BLOCK_ALGORITHM =
+    AgentReducePolicy::BLOCK_ALGORITHM;
+
+  /// Parameterized BlockReduce primitive
+  using BlockReduceT =
+    BlockReduce<AccumT, BLOCK_THREADS, AgentReducePolicy::BLOCK_ALGORITHM>;
+
+  /// Shared memory type required by this thread block
+  struct _TempStorage
+  {
+    typename BlockReduceT::TempStorage reduce;
+  };
+
+  /// Alias wrapper allowing storage to be unioned
+  struct TempStorage : Uninitialized<_TempStorage>
+  {};
+
+  //---------------------------------------------------------------------
+  // Per-thread fields
+  //---------------------------------------------------------------------
+
+  _TempStorage &temp_storage;         ///< Reference to temp_storage
+  InputIteratorT d_in;                ///< Input data to reduce
+  WrappedInputIteratorT d_wrapped_in; ///< Wrapped input data to reduce
+  ReductionOp reduction_op;           ///< Binary reduction operator
+
+  //---------------------------------------------------------------------
+  // Utility
+  //---------------------------------------------------------------------
+
+  // Whether or not the input is aligned with the vector type (specialized for
+  // types we can vectorize)
+  template <typename Iterator>
+  static __device__ __forceinline__ bool
+  IsAligned(Iterator d_in, Int2Type<true> /*can_vectorize*/)
+  {
+    return (size_t(d_in) & (sizeof(VectorT) - 1)) == 0;
+  }
+
+  // Whether or not the input is aligned with the vector type (specialized for
+  // types we cannot vectorize)
+  template <typename Iterator>
+  static __device__ __forceinline__ bool
+  IsAligned(Iterator /*d_in*/, Int2Type<false> /*can_vectorize*/)
+  {
+    return false;
+  }
+
+  //---------------------------------------------------------------------
+  // Constructor
+  //---------------------------------------------------------------------
+
+  /**
+   * @brief Constructor
+   * @param temp_storage Reference to temp_storage
+   * @param d_in Input data to reduce
+   * @param reduction_op Binary reduction operator
+   */
+  __device__ __forceinline__ AgentReduce(TempStorage &temp_storage,
+                                         InputIteratorT d_in,
+                                         ReductionOp reduction_op)
+      : temp_storage(temp_storage.Alias())
+      , d_in(d_in)
+      , d_wrapped_in(d_in)
+      , reduction_op(reduction_op)
+  {}
+
+  //---------------------------------------------------------------------
+  // Tile consumption
+  //---------------------------------------------------------------------
+
+  /**
+   * @brief Consume a full tile of input (non-vectorized)
+   * @param block_offset The offset the tile to consume
```

Comment:
We are missing the `thread_aggregate` param or is that conventionally skipped?

Reply:
Good catch. I guess that only input parameters are documented here. Before refactoring, this was undocumented as well. I hope that we'll be able to rewrite this to return `thread_aggreate` instead of taking it as a parameter.